File: src/main/java/xyz/devvydont/smprpg/enchantments/definitions/IgnoranceBlessing.java
package xyz.devvydont.smprpg.enchantments.definitions;

import io.papermc.paper.registry.RegistryKey;
import io.papermc.paper.registry.keys.tags.ItemTypeTagKeys;
import io.papermc.paper.registry.set.RegistryKeySet;
import io.papermc.paper.registry.set.RegistrySet;
import io.papermc.paper.registry.tag.TagKey;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextColor;
import org.bukkit.Bukkit;
import org.bukkit.Sound;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockDropItemEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.inventory.EquipmentSlotGroup;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ItemType;
import org.jetbrains.annotations.NotNull;
import xyz.devvydont.smprpg.SMPRPG;
import xyz.devvydont.smprpg.ability.AbilityCost;
import xyz.devvydont.smprpg.enchantments.CustomEnchantment;
import xyz.devvydont.smprpg.enchantments.EnchantmentRarity;
import xyz.devvydont.smprpg.events.abilities.AbilityCastEvent;
import xyz.devvydont.smprpg.services.ActionBarService;
import xyz.devvydont.smprpg.services.DropsService;
import xyz.devvydont.smprpg.services.EnchantmentService;
import xyz.devvydont.smprpg.util.formatting.ComponentUtils;

import java.util.Map;
import java.util.UUID;

public class IgnoranceBlessing extends CustomEnchantment implements Listener {

    public IgnoranceBlessing(String id) {
        super(id);
    }

    public static int getManaCostReduction(int level) {
        return level * 10;
    }

    @Override
    public @NotNull Component getDisplayName() {
        return ComponentUtils.create("Blessing of Ignorance", NamedTextColor.YELLOW);
    }

    @Override
    public @NotNull TextColor getEnchantColor() {
        return NamedTextColor.YELLOW;
    }

    @Override
    public @NotNull Component getDescription() {
        return ComponentUtils.merge(
            ComponentUtils.create("Reduces the cost of "),
            ComponentUtils.create("mana based attacks and abilities", NamedTextColor.AQUA),
            ComponentUtils.create(" by " + getManaCostReduction(getLevel()) + "%")
        );
    }

    @Override
    public TagKey<ItemType> getItemTypeTag() {
        return ItemTypeTagKeys.ENCHANTABLE_WEAPON;
    }

    @Override
    public int getAnvilCost() {
        return 1;
    }

    @Override
    public int getMaxLevel() {
        return 5;
    }

    @Override
    public int getWeight() {
        return EnchantmentRarity.BLESSING.getWeight();
    }

    @Override
    public EquipmentSlotGroup getEquipmentSlotGroup() {
        return EquipmentSlotGroup.HEAD;
    }

    @Override
    public int getSkillRequirement() {
        return 25;
    }

    @Override
    public @NotNull RegistryKeySet<Enchantment> getConflictingEnchantments() {
        return RegistrySet.keySet(RegistryKey.ENCHANTMENT,
                EnchantmentService.KEEPING_BLESSING.getTypedKey(),
                EnchantmentService.MERCY_BLESSING.getTypedKey(),
                EnchantmentService.VOIDSTRIDING_BLESSING.getTypedKey(),
                EnchantmentService.REPLENISHING.getTypedKey(),
                EnchantmentService.TELEKINESIS_BLESSING.getTypedKey()
        );
    }

    @EventHandler
    private void __onAbilityCast(AbilityCastEvent event){
        if (event.getAbilityCost().resource == AbilityCost.Resource.MANA) {
            var ench = event.getItem().getEnchantmentLevel(this.getEnchantment());
            if (ench > 0)
                event.setAbilityCost(event.getAbilityCost().reduce(getManaCostReduction(ench) / 100.0));
        }
    }
}

File: src/main/java/xyz/devvydont/smprpg/enchantments/definitions/IgnoranceBlessing.kt
package xyz.devvydont.smprpg.enchantments.definitions

import io.papermc.paper.registry.RegistryKey
import io.papermc.paper.registry.keys.tags.ItemTypeTagKeys
import io.papermc.paper.registry.set.RegistryKeySet
import io.papermc.paper.registry.set.RegistrySet
import io.papermc.paper.registry.tag.TagKey
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextColor
import org.bukkit.Color
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.inventory.EquipmentSlotGroup
import org.bukkit.inventory.ItemType
import xyz.devvydont.smprpg.SMPRPG
import xyz.devvydont.smprpg.ability.AbilityCost
import xyz.devvydont.smprpg.enchantments.CustomEnchantment
import xyz.devvydont.smprpg.enchantments.EnchantmentRarity
import xyz.devvydont.smprpg.enchantments.ScrollColor
import xyz.devvydont.smprpg.events.abilities.AbilityCastEvent
import xyz.devvydont.smprpg.services.EnchantmentService
import xyz.devvydont.smprpg.services.EntityService
import xyz.devvydont.smprpg.util.formatting.ComponentUtils

class IgnoranceBlessing(id: String) : CustomEnchantment(id), Listener {
    override val displayName: Component get() = ComponentUtils.create("Blessing of Ignorance", NamedTextColor.YELLOW)
    override val description: Component get() = ComponentUtils.merge(
            ComponentUtils.create("Reduces the cost of "),
            ComponentUtils.create("mana based attacks and abilities", NamedTextColor.AQUA),
            ComponentUtils.create(" by " + getManaCostReduction(level) + "%")
        )
    override val enchantColor: TextColor get() = NamedTextColor.YELLOW
    override val scrollColor: Color get()        = ScrollColor.BLESSING.color
    override val scrollBindingColor: Color get() = Color.fromRGB(0, 0, 255)

    override val itemTypeTag: TagKey<ItemType> get()           = ItemTypeTagKeys.ENCHANTABLE_WEAPON
    override val maxLevel: Int get()                           = 5
    override val weight: Int get()                             = EnchantmentRarity.BLESSING.weight
    override val isBlessing: Boolean get()                     = true
    override val equipmentSlotGroup: EquipmentSlotGroup? get() = EquipmentSlotGroup.ANY
    override val skillRequirement: Int get()                   = 25

    override val conflictingEnchantments: RegistryKeySet<Enchantment>
        get() = RegistrySet.keySet(
            RegistryKey.ENCHANTMENT,
            EnchantmentService.KEEPING_BLESSING.typedKey,
            EnchantmentService.MERCY_BLESSING.typedKey,
            EnchantmentService.VOIDSTRIDING_BLESSING.typedKey,
            EnchantmentService.REPLENISHING.typedKey,
            EnchantmentService.TELEKINESIS_BLESSING.typedKey
        )

    @EventHandler
    private fun onAbilityCast(event: AbilityCastEvent) {
        val dealer = event.player
        if (!isEnchantmentActive(dealer.player.equipment.itemInMainHand, dealer)) return

        if (event.abilityCost.resource == AbilityCost.Resource.MANA) {
            val ench = event.item.getEnchantmentLevel(this.enchantment)
            if (ench > 0) event.abilityCost = event.abilityCost.reduce(getManaCostReduction(ench) / 100.0)
        }
    }

    companion object {
        fun getManaCostReduction(level: Int): Int { return level * 10 }
    }
}
